sql: CANCEL QUERIES (and SESSINOS) doesn't unwrap types properly

Found with RSG (cc @mjibson)

```
panic: interface conversion: tree.Datum is *tree.DOidWrapper, not *tree.DString [recovered]
	panic: panic while executing 1 statements: CANCEL QUERIES IF EXISTS VALUES (_::NAME); caused by interface conversion: tree.Datum is *tree.DOidWrapper, not *tree.DString
```

--- a/pkg/sql/cancel_queries.go
+++ b/pkg/sql/cancel_queries.go
@@ -21,6 +21,7 @@ import (
 	"github.com/pkg/errors"
 
 	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/types"
 )
@@ -64,12 +65,15 @@ func (n *cancelQueriesNode) Next(params runParams) (bool, error) {
 	}
 
 	statusServer := params.extendedEvalCtx.StatusServer
-	queryIDDatum := datum.(*tree.DString)
+	queryIDString, ok := tree.AsDString(datum)
+	if !ok {
+		return false, pgerror.NewErrorf(pgerror.CodeInternalError,
+			"programming error: %q: expected *DString, found %T", datum, datum)
+	}
 
-	queryIDString := string(*queryIDDatum)
-	queryID, err := StringToClusterWideID(queryIDString)
+	queryID, err := StringToClusterWideID(string(queryIDString))
 	if err != nil {
-		return false, errors.Wrapf(err, "invalid query ID '%s'", queryIDString)
+		return false, errors.Wrapf(err, "invalid query ID %s", datum)
 	}
 
 	// Get the lowest 32 bits of the query ID.
@@ -77,7 +81,7 @@ func (n *cancelQueriesNode) Next(params runParams) (bool, error) {
 
 	request := &serverpb.CancelQueryRequest{
 		NodeId:   fmt.Sprintf("%d", nodeID),
-		QueryID:  queryIDString,
+		QueryID:  string(queryIDString),
 		Username: params.SessionData().User,
 	}
 

--- a/pkg/sql/cancel_sessions.go
+++ b/pkg/sql/cancel_sessions.go
@@ -19,6 +19,7 @@ import (
 	"fmt"
 
 	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/types"
 	"github.com/pkg/errors"
@@ -63,12 +64,15 @@ func (n *cancelSessionsNode) Next(params runParams) (bool, error) {
 	}
 
 	statusServer := params.extendedEvalCtx.StatusServer
-	sessionIDDatum := datum.(*tree.DString)
+	sessionIDString, ok := tree.AsDString(datum)
+	if !ok {
+		return false, pgerror.NewErrorf(pgerror.CodeInternalError,
+			"programming error: %q: expected *DString, found %T", datum, datum)
+	}
 
-	sessionIDString := string(*sessionIDDatum)
-	sessionID, err := StringToClusterWideID(sessionIDString)
+	sessionID, err := StringToClusterWideID(string(sessionIDString))
 	if err != nil {
-		return false, errors.Wrapf(err, "invalid session ID '%s'", sessionIDString)
+		return false, errors.Wrapf(err, "invalid session ID %s", datum)
 	}
 
 	// Get the lowest 32 bits of the session ID.

--- a/pkg/sql/control_jobs.go
+++ b/pkg/sql/control_jobs.go
@@ -16,9 +16,9 @@ package sql
 
 import (
 	"context"
-	"fmt"
 
 	"github.com/cockroachdb/cockroach/pkg/sql/jobs"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/types"
 	"github.com/pkg/errors"
@@ -47,7 +47,7 @@ func (p *planner) ControlJobs(ctx context.Context, n *tree.ControlJobs) (planNod
 			tree.JobCommandToStatement[n.Command], len(cols))
 	}
 	if !cols[0].Typ.Equivalent(types.Int) {
-		return nil, errors.Errorf("%s QUERIES requires int values, not type %s",
+		return nil, errors.Errorf("%s JOBS requires int values, not type %s",
 			tree.JobCommandToStatement[n.Command], cols[0].Typ)
 	}
 
@@ -74,15 +74,15 @@ func (n *controlJobsNode) startExec(params runParams) error {
 			break
 		}
 
-		datum := n.rows.Values()[0]
-		if datum == tree.DNull {
+		jobIDDatum := n.rows.Values()[0]
+		if jobIDDatum == tree.DNull {
 			continue
 		}
 
-		jobIDDatum := datum.(*tree.DInt)
 		jobID, ok := tree.AsDInt(jobIDDatum)
 		if !ok {
-			return fmt.Errorf("%s is not a valid job ID", jobIDDatum)
+			return pgerror.NewErrorf(pgerror.CodeInternalError,
+				"programming error: %q: expected *DInt, found %T", jobIDDatum, jobIDDatum)
 		}
 
 		switch n.desiredStatus {
@@ -93,7 +93,8 @@ func (n *controlJobsNode) startExec(params runParams) error {
 		case jobs.StatusCanceled:
 			err = reg.Cancel(params.ctx, params.p.txn, int64(jobID))
 		default:
-			err = fmt.Errorf("programmer error: unhandled status %v", n.desiredStatus)
+			err = pgerror.NewErrorf(pgerror.CodeInternalError,
+				"programming error: unhandled status %v", n.desiredStatus)
 		}
 		if err != nil {
 			return err

--- a/pkg/sql/logictest/testdata/logic_test/run_control
+++ b/pkg/sql/logictest/testdata/logic_test/run_control
@@ -12,6 +12,9 @@ CANCEL JOBS SELECT 'foo'
 query error CANCEL JOBS expects a single column source, got 2 columns
 CANCEL JOBS VALUES (1,2)
 
+query error CANCEL JOBS requires int values, not type oid
+CANCEL JOB 1::OID
+
 statement ok count 0
 PAUSE JOB (SELECT id FROM system.jobs LIMIT 0)
 
@@ -83,3 +86,10 @@ CANCEL SESSION (SELECT 'a' LIMIT 0)
 
 statement ok count 0
 CANCEL SESSIONS SELECT 'a' LIMIT 0
+
+# Regression test for #25842
+query error odd length hex string
+CANCEL SESSION 'aaa'::NAME
+
+query error odd length hex string
+CANCEL QUERY 'aaa'::NAME

